dill/_dill.py

```diff
@@ -410,9 +416,69 @@ def save_numpy_array(pickler, obj):
             raise PicklingError(msg)
         logger.trace_setup(self)
         StockPickler.dump(self, obj)
-
     dump.__doc__ = StockPickler.dump.__doc__
 
+    def save(self, obj, save_persistent_id=True, *, name=None):
+        """If self._refonfail is True, try to save object by reference if pickling fails."""
+        if not self._refonfail:
+            super().save(obj, save_persistent_id)
+            return
+        # Disable framing (right after the framer.init_framing() call at dump()).
+        self.framer.current_frame = None
+        # Store initial state.
+        position = self._file_tell()
+        memo_size = len(self.memo)
+        try:
+            super().save(obj, save_persistent_id)
+        except UNPICKLEABLE_ERRORS + (AttributeError,) as error_stack:
+            # AttributeError may happen in the save_global() call from a child object.
+            if (type(error_stack) == AttributeError
+                    and "no attribute '__name__'" not in error_stack.args[0]):
+                raise
+            # Roll back the stream.
+            self._file_seek(position)
+            self._file_truncate()
+            # Roll back memo.
+            for _ in range(len(self.memo) - memo_size):
+                self.memo.popitem()  # LIFO order is guaranteed since 3.7
+            try:
+                if isinstance(obj, ModuleType) and \
+                        (_is_builtin_module(obj) or obj is sys.modules['dill']):
+                    self.save_reduce(_import_module, (obj.__name__,), obj=obj)
+                else:
+                    self.save_global(obj, name)
+            except (AttributeError, PicklingError) as error:
+                if getattr(self, '_trace_stack', None) and id(obj) == self._trace_stack[-1]:
+                    # Roll back trace state.
+                    self._trace_stack.pop()
+                    self._size_stack.pop()
+                raise error from error_stack
+            logger.trace(self, "# X: fallback to save_global: <%s object at %#012x>",
+                         type(obj).__name__, id(obj), obj=obj)
+
+    def _save_module_dict(self, obj):
+        """
+        Use object name in the module namespace as a last resource to try to
+        save it by reference when pickling fails.
+
+        Modified from Pickler.save_dict() and Pickler._batch_setitems().
```

Comment:
This should be in a comment (not in the docstring), and may need a reference to the license (depending on what the license says).

Reply:
I followed the Python License requirements and added the License Agreement in `LICENSE`, plus a copyright notice and a brief summary of the modifications to the code in the source file. Just pushed this, take a look.

Reply:
> PSF's License Agreement and PSF's notice of copyright, i.e., "Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006 Python Software Foundation; All Rights Reserved" are retained in Python alone or in any derivative version prepared by Licensee.

> In the event Licensee prepares a derivative work that is based on or incorporates Python or any part thereof, and wants to make the derivative work available to others as provided herein, then Licensee hereby agrees to include in any such work a brief summary of the changes made to Python.

It should be sufficient to summarize your changes, tag it with the copyright (right where you made the changes) and a link to the python license.

Reply:
Copyright law is hairy, but what I understood after reading a bit is that "derivative work" for code means copying, with or without changes, "major components" of the software. I believe that half a dozen lines of a function that must match the class API and the pickle protocol (and therefore can't be much different of what it is) don't constitute a major component. In this sense it wouldn't be necessary to add the copyright notice. But giving credit doesn't hurt.